src/coreclr/jit/importercalls.cpp

```diff
@@ -5199,8 +5191,18 @@ GenTree* Compiler::impPrimitiveNamedIntrinsic(NamedIntrinsic        intrinsic,
 
     switch (intrinsic)
     {
-        case NI_PRIMITIVE_ConvertToInteger:
         case NI_PRIMITIVE_ConvertToIntegerNative:
+        {
+            if (opts.IsReadyToRun())
+            {
+                // We explicitly block these APIs from being expanded in R2R
+                // since we know they are non-deterministic across hardware
+                return nullptr;
+            }
```

Comment:
```suggestion
            if (BlockNonDeterministicIntrinsics(mustExpand))
            {
                return nullptr;
            }
```
Create a helper function for this. 

```
bool BlockNonDeterministicIntrinsics(bool mustExpand)
{
    // We explicitly block these APIs from being expanded in R2R
    // since we know they are non-deterministic across hardware
    if (opts.IsReadyToRun() && !compiler->IsTargetAbi(CORINFO_NATIVEAOT_ABI))
    {
        if (mustExpand)
            implLimitation();
        return true;
    }
    return false;
}
```
- I think this needs to exclude NAOT. We do not want to block this expansion for NAOT. R2R is true for NativeAOT: https://github.com/dotnet/runtime/blob/bdfbed6de953957fd4dbba9c8cba0a3c47dde1f8/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs#L4122 . 
- This can handle both the inlining expansion and self-recursive expansion. It will save us from having to duplicate this logic in CoreLib and trying the two it in sync.

Reply:
> I think this needs to exclude NAOT. We do not want to block this expansion for NAOT. R2R is true for NativeAOT

Ah, I see. 

Notably I think this means there are be a few spots we're using `IsReadyToRun` "incorrectly" throughout the JIT. For example, some places in valuenum that are avoiding constant folding based on this when it shouldn't since NAOT knows the exact target ISA set.

Not something to handle here, but maybe something we should minimally look into to improve codegen for NAOT as a separate topic

Reply:
Updated the PR to:
* have the helper function given above
* remove the `BypassReadyToRun` attribute from source
* ensure the managed implementations are recursive where relevant
* have an explicit regression test covering the APIs and comparing the indirect/direct invocations for known edge case values
* add a section to the botr page on vectors-and-intrinsics to account for non-deterministic intrinsic handling